// Generated by https://github.com/foxglove/schemas

syntax = "proto3";

import "Pose.proto";
import "google/protobuf/timestamp.proto";
import "header.proto";
import "Vector3.proto";

package foxglove;

message RobotLinkInfo {
  string name = 1;
  foxglove.Pose transform = 2;
}

message RobotJointInfo {
  string name = 1;
  double value = 2;
}

message PathPoint {
  double x = 1;
  double y = 2;
  double z = 3;
  double yaw = 4;
}

message WheelVelocities {
  double left_wheel = 1;
  double right_wheel = 2;
}

message RobotMotion {
  double linear_velocity = 1;
  double angular_velocity = 2;
  double steering_angle = 3;
}

message MPCData {
  repeated foxglove.Vector3 global_path = 1;
  repeated foxglove.Vector3 mpc_prediction = 2;
  repeated foxglove.Vector3 current_reference = 3;
  PathPoint start_point = 4;
  PathPoint end_point = 5;
  repeated foxglove.Vector3 real_time_path = 6;
  WheelVelocities wheel_velocities = 7;
  RobotMotion robot_motion = 8;
  string mpc_status = 9;
  double timestamp = 10;
}

enum BaselinkType {
    BASELINK_TYPE_UNSPECIFIED = 0;  // 默认值（proto3 要求 enum 从 0 开始）
    USE_BASELINK_AS_POSE = 1;      // 使用 base_link 作为机器人位姿
    IGNORE_BASELINK = 2;           // 忽略 base_link
}

message RobotInfo {
  google.protobuf.Timestamp timestamp = 1;

  string frame_id = 2;

  string name = 3;

  foxglove.Pose world_pose = 4;

  repeated RobotLinkInfo links = 5;

  arnold.common.proto.Header header = 6;

  repeated RobotJointInfo joints = 7;

  RobotLinkInfo base_link = 8;

  BaselinkType baselink_type = 9;

  RobotLinkInfo gripper_end_effort_link = 10;

  MPCData mpc_data = 11;
}
